[Structural] Compilation error in master branch using Clang 7.0.1 (macOS)

Today I tried to compile the StructuralMechanics Application in my mac (after some weeks of not doing so), and got the following compilation error:

```
Scanning dependencies of target KratosStructuralMechanicsCore
[ 46%] Building CXX object applications/StructuralMechanicsApplication/CMakeFiles/KratosStructuralMechanicsCore.dir/structural_mechanics_application.cpp.o
[ 46%] Building CXX object applications/SolidMechanicsApplication/CMakeFiles/KratosSolidMechanicsCore.dir/custom_elements/beam_elements/geometrically_exact_rod_element.cpp.o
In file included from /Users/ipouplana/Kratos/applications/StructuralMechanicsApplication/structural_mechanics_application.cpp:20:
In file included from /Users/ipouplana/Kratos/applications/StructuralMechanicsApplication/structural_mechanics_application_variables.h:22:
In file included from /Users/ipouplana/Kratos/applications/StructuralMechanicsApplication/custom_utilities/shell_cross_section.hpp:23:
/Users/ipouplana/Kratos/kratos/includes/serializer.h:549:17: error: no member
      named 'save' in 'std::__1::__bit_const_reference<std::__1::vector<bool,
      std::__1::allocator<bool> > >'
        rObject.save(*this);
        ~~~~~~~ ^
/Users/ipouplana/Kratos/kratos/includes/serializer.h:502:13: note: in
      instantiation of function template specialization
      'Kratos::Serializer::save<std::__1::__bit_const_reference<std::__1::vector<bool,
      std::__1::allocator<bool> > > >' requested here
            save("E", rObject[i]);
            ^
/Users/ipouplana/Kratos/applications/StructuralMechanicsApplication/custom_constitutive/truss_plasticity_constitutive_law.h:252:21: note: 
      in instantiation of function template specialization
      'Kratos::Serializer::save<bool>' requested here
        rSerializer.save("InelasticFlag", this->mInElasticFlagVector);
                    ^
1 error generated.
make[2]: *** [applications/StructuralMechanicsApplication/CMakeFiles/KratosStructuralMechanicsCore.dir/structural_mechanics_application.cpp.o] Error 1
make[1]: *** [applications/StructuralMechanicsApplication/CMakeFiles/KratosStructuralMechanicsCore.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

I got this error in the current version of the 'master' branch (commit 6400624e867cbf76b7268f3a4abdabc1f2d87eff) and I have the following configuration:

```
-- The C compiler identification is Clang 7.0.1
-- The CXX compiler identification is Clang 7.0.1
-- Check for working C compiler: /usr/local/opt/llvm/bin/clang
-- Check for working C compiler: /usr/local/opt/llvm/bin/clang -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/local/opt/llvm/bin/clang++
-- Check for working CXX compiler: /usr/local/opt/llvm/bin/clang++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Deprecation Warning at CMakeLists.txt:8 (cmake_policy):
  The OLD behavior for policy CMP0054 will be removed from a future version
  of CMake.

  The cmake-policies(7) manual explains that the OLD behaviors of all
  policies are deprecated and that a policy should be set to OLD only under
  specific short-term circumstances.  Projects should be ported to the NEW
  behavior and not rely on setting a policy to OLD.


-- Found Git: /usr/bin/git (found version "2.17.2 (Apple Git-113)") 
KRATOS_BUILD_TESTING not defined. Setting to ON
-- Found PythonInterp: /Library/Frameworks/Python.framework/Versions/3.6/bin/python3 (found version "3.6.6") 
-- Found PythonLibs: /Library/Frameworks/Python.framework/Versions/3.6/lib/libpython3.6m.dylib
-- Performing Test HAS_CPP14_FLAG
-- Performing Test HAS_CPP14_FLAG - Success
-- OpenMP included
-- Found OpenMP_C: -fopenmp=libomp (found version "3.1") 
-- Found OpenMP_CXX: -fopenmp=libomp (found version "3.1") 
-- Found OpenMP: TRUE (found version "3.1")  
-- Boost version: 1.67.0
-- Boost Include: /Users/ipouplana/boost_1_67_0
-- Boost Linkdir: 
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - found
-- Found Threads: TRUE  
-- Found BLAS: /usr/lib/libblas.dylib  
linker flags with blas =
BLA_VENDOR = All
-- Found BLAS: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib  
-- A library with LAPACK API found.
LAPACK = /usr/lib/liblapack.dylib
LAPACK modified = /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libLAPACK.dylib
-- Found ZLIB: /usr/lib/libz.dylib (found version "1.2.11") 
-- AMatrix not found. Please define the AMATRIX_DIR pointing to the AMatrix root directory.
 
CMAKE_BUILD_TYPE ........ Release
 
WARNING: YOUR GIDPOST IS BEING COMPILED  WITHOUT DEBUG (ALL ASSERTS WILL BE AVOIDED)
-- Performing Test HAS_FLTO_THIN
-- Performing Test HAS_FLTO_THIN - Success
-- LTO enabled
 
 
***** LIST OF KRATOS APPLICATIONS THAT ARE GOING TO BE COMPILED: *****
 
                CONVECTION_DIFFUSION_APPLICATION - ON
                STRUCTURAL_MECHANICS_APPLICATION - ON
                SOLID_MECHANICS_APPLICATION - ON
                CONSTITUTIVE_MODELS_APPLICATION - ON
                POROMECHANICS_APPLICATION - ON
                DAM_APPLICATION - ON
                FLUID_DYNAMICS_APPLICATION - ON
                EXTERNAL_SOLVERS_APPLICATION - ON
 
**********************************************************************
 
****** LIST OF KRATOS APPLICATIONS THAT ARE SET TO OFF: **************
 
                MESH_MOVING_APPLICATION - OFF
                FSI_APPLICATION - OFF
                MIXED_ELEMENT_APPLICATION - OFF
                PFEM_FLUID_DYNAMICS_APPLICATION - OFF
                DEM_APPLICATION - OFF
                SWIMMING_DEM_APPLICATION - OFF
                TRILINOS_APPLICATION - OFF
                METIS_APPLICATION - OFF
                MESHING_APPLICATION - OFF
                SHAPE_OPTIMIZATION_APPLICATION - OFF
                TOPOLOGY_OPTIMIZATION_APPLICATION - OFF
                DELAUNAY_MESHING_APPLICATION - OFF
 
**********************************************************************
 
 
**** Configuring KratosExternalSolversApplication ****
**** configuring KratosConstitutiveModelsApplication ****
**** configuring KratosSolidMechanicsApplication ****
**** configuring KratosConvectionDiffusionApplication ****
**** configuring KratosStructuralMechanicsApplication ****
**** configuring KratosFluidDynamicsApplication ****
**** configuring KratosPoromechanicsApplication ****
**** configuring KratosDamApplication ****
-- triangle : /Users/ipouplana/Kratos/external_libraries/triangle
python libraries to be installed are: /Library/Frameworks/Python.framework/Versions/3.6/Python
-- Standard install dir /Users/ipouplana/Kratos
installed blas = /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
installed lapack = /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libLAPACK.dylib
-- Configuring done
```

The strange thing is that I did not get this error when compiling the same code using GCC 7.3.0 in Ubuntu 18.04...

Maybe I'm doing something wrong, but I don't know what is happening. Does someone have a clue on why I get the compilation error ? 

Thanks in advance.

--- a/applications/StructuralMechanicsApplication/custom_constitutive/truss_plasticity_constitutive_law.cpp
+++ b/applications/StructuralMechanicsApplication/custom_constitutive/truss_plasticity_constitutive_law.cpp
@@ -72,7 +72,7 @@ bool& TrussPlasticityConstitutiveLaw::GetValue(
     bool& rValue
     )
 {
-    if(rThisVariable == INELASTIC_FLAG) rValue = this->mInElasticFlagVector[0];
+    if(rThisVariable == INELASTIC_FLAG) rValue = this->mCurrentInElasticFlag;
     else KRATOS_ERROR << "Can't get the specified value" << std::endl;
     return rValue;
 }
@@ -123,7 +123,7 @@ void TrussPlasticityConstitutiveLaw::SetValue(const Variable<bool>& rThisVariabl
                           const bool& rValue,
                           const ProcessInfo& rCurrentProcessInfo)
 {
-    if(rThisVariable == INELASTIC_FLAG) this->mInElasticFlagVector[0] = rValue;
+    if(rThisVariable == INELASTIC_FLAG) this->mCurrentInElasticFlag = rValue;
     else KRATOS_ERROR << "Can't set the specified value" << std::endl;
 }
 
@@ -143,7 +143,7 @@ double& TrussPlasticityConstitutiveLaw::CalculateValue(
         const Properties& r_material_properties = rParameterValues.GetMaterialProperties();
         const double hardening_modulus = r_material_properties[HARDENING_MODULUS_1D];
         const double youngs_modulus = r_material_properties[YOUNG_MODULUS];
-        if (this->mInElasticFlagVector[0])
+        if (this->mCurrentInElasticFlag)
         {
             KRATOS_DEBUG_ERROR_IF((hardening_modulus+youngs_modulus)<numerical_limit)
              << "Dividing by 0 when calculating the plastic tangent modulus" << std::endl;
@@ -172,7 +172,7 @@ Vector& TrussPlasticityConstitutiveLaw::CalculateValue(
         double current_stress = this->mStressState;
 
 
-        if (this->mInElasticFlagVector[0])
+        if (this->mCurrentInElasticFlag)
         {
             const double hardening_modulus = r_material_properties[HARDENING_MODULUS_1D];
             const double youngs_modulus = r_material_properties[YOUNG_MODULUS];
@@ -241,7 +241,7 @@ void TrussPlasticityConstitutiveLaw::CalculateMaterialResponse(
     if (SaveInternalVariables)
     {
         this->mStressState = rStressVector[0];
-        this->mInElasticFlagVector[0] = this->CheckIfIsPlasticRegime(rMaterialProperties,rStressVector[0]);
+        this->mCurrentInElasticFlag = this->CheckIfIsPlasticRegime(rMaterialProperties,rStressVector[0]);
 
         if (this->CheckPlasticIterationHistory())
         {
@@ -320,7 +320,7 @@ void TrussPlasticityConstitutiveLaw::FinalizeNonLinearIteration(const Properties
                     const Vector& rShapeFunctionsValues,
                     const ProcessInfo& rCurrentProcessInfo)
 {
-    this->mInElasticFlagVector[1] = this->mInElasticFlagVector[0];
+    this->mPreviousInElasticFlag = this->mCurrentInElasticFlag;
 }
 
 void TrussPlasticityConstitutiveLaw::FinalizeSolutionStep(const Properties& rMaterialProperties,
@@ -330,8 +330,8 @@ void TrussPlasticityConstitutiveLaw::FinalizeSolutionStep(const Properties& rMat
 {
     this->mAccumulatedPlasticStrainVector[1] = this->mAccumulatedPlasticStrainVector[0];
     this->mPlasticAlphaVector[1] = this->mPlasticAlphaVector[0];
-    this->mInElasticFlagVector[0] = false;
-    this->mInElasticFlagVector[1] = false;
+    this->mCurrentInElasticFlag = false;
+    this->mPreviousInElasticFlag = false;
 }
 
 } // Namespace Kratos

--- a/applications/StructuralMechanicsApplication/custom_constitutive/truss_plasticity_constitutive_law.h
+++ b/applications/StructuralMechanicsApplication/custom_constitutive/truss_plasticity_constitutive_law.h
@@ -157,7 +157,7 @@ public:
     bool CheckPlasticIterationHistory() const
     {
         bool check_flag = false;
-        if(this->mInElasticFlagVector[1] && !this->mInElasticFlagVector[0])
+        if(this->mPreviousInElasticFlag && !this->mCurrentInElasticFlag)
         {
             check_flag = true;
         }
@@ -215,10 +215,13 @@ private:
     ///@{
     double mStressState = 0.0; // The current stress state
 
-    //the following members are vectors where
-    //[0] is the current non_linear iteration step
-    //[1] is the previous non_linear iteration step
-    std::vector<bool> mInElasticFlagVector = {false, false}; /// This flags tells if we are in a elastic or ineslastic regime
+    //the following members are
+    //mCurrentInElasticFlag is the current non_linear iteration step
+    //mPreviousInElasticFlag is the previous non_linear iteration step
+    bool mPreviousInElasticFlag = false;/// This flags tells if we are in a elastic or ineslastic regime
+    bool mCurrentInElasticFlag = false;/// This flags tells if we are in a elastic or ineslastic regime
+
+
     BoundedVector<double, 2> mPlasticAlphaVector = ZeroVector(2); /// The current plastic increment
     BoundedVector<double, 2> mAccumulatedPlasticStrainVector = ZeroVector(2); /// The current accumulated plastic strain
 
@@ -249,7 +252,8 @@ private:
         rSerializer.save("StressState", this->mStressState);
         rSerializer.save("PlasticAlpha", this->mPlasticAlphaVector);
         rSerializer.save("AccumulatedPlasticStrain", this->mAccumulatedPlasticStrainVector);
-        rSerializer.save("InelasticFlag", this->mInElasticFlagVector);
+        rSerializer.save("PreviousInelasticFlag", this->mPreviousInElasticFlag);
+        rSerializer.save("CurrentInElasticFlag", this->mCurrentInElasticFlag);
     }
 
     void load(Serializer& rSerializer) override
@@ -258,7 +262,8 @@ private:
         rSerializer.load("StressState", this->mStressState);
         rSerializer.load("PlasticAlpha", this->mPlasticAlphaVector);
         rSerializer.load("AccumulatedPlasticStrain", this->mAccumulatedPlasticStrainVector);
-        rSerializer.load("InelasticFlag", this->mInElasticFlagVector);
+        rSerializer.load("PreviousInelasticFlag", this->mPreviousInElasticFlag);
+        rSerializer.load("CurrentInElasticFlag", this->mCurrentInElasticFlag);
     }
 
 

